### ISSUE TITLE
Running the program on a bike

### ISSUE THREAD

**Issue by JohannesAppel** [JohannesAppel/johannesappel-#12]
@WK-GiHu
Now that the program is running it needs to be able to read the data from a bike.
Up until now it ran on a simulated box, which only outputs 4 lines of data. 
There are a few challenges with the bike:

1.  Data isn't read at the same speed. Some of the data is only read every 500ms, where other are read all the time.
2.  With no.1 the amount of data is inconsistent every time you read. So we need to update the data that changed and rewrite everything.
3.  There is an unknown number of data lines. That's why I started with the `counter` function.

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>3. There is an unknown number of data lines.

Start with point 3., enable 
https://github.com/JohannesAppel/johannesappel-/blob/b99b1aba8136c3e543465aac0f04ba4d1c3c2017/DataLogger/src/canbus.py#L121

and get it running in `canbus.py __main__`.
If you get valid detection of the lines move `.counter(...` into `.__init__` and the `tkinter` part will use allready.

---

>2. With no.1 the amount of data is inconsistent every time you read.

Here i need a example output that show the differences.

**Comment by JohannesAppel** [comment]
@WK-GiHu 
I think I might be doing something wrong. To where in `.__init__` should I move `.counter(...`, can't get it to work again.

>Here i need a example output that show the differences.

I've attached screenshots of the outputs at various times.
![Screenshot (26)](https://user-images.githubusercontent.com/58848423/71873640-5d2e7500-3128-11ea-8744-979f6eace03b.png)
![Screenshot (24)](https://user-images.githubusercontent.com/58848423/71873661-6881a080-3128-11ea-88d5-a01cbc0c4057.png)
![Screenshot (23)](https://user-images.githubusercontent.com/58848423/71873685-79321680-3128-11ea-8aec-a14ba49b275e.png)
![Screenshot (25)](https://user-images.githubusercontent.com/58848423/71873720-8bac5000-3128-11ea-8dc8-82020b457032.png)

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>To where in .__init__ should I move .counter(..., can't get it to work again.

Copy Line 21
https://github.com/JohannesAppel/johannesappel-/blob/f3b4059698cf6997470277693270a2ebbf220270/DataLogger/src/canbus.py#L121
to Line 23
https://github.com/JohannesAppel/johannesappel-/blob/f3b4059698cf6997470277693270a2ebbf220270/DataLogger/src/canbus.py#L23
and rename to
```
self.cnt = cbus.counter()
```

---

>>With no.1 the amount of data is inconsistent every time you read.
>Here i need a example output that show the differences.

You mean the different length of the `bytearray(...`?
The Frame 304 is always shorter?

**Comment by JohannesAppel** [comment]
@WK-GiHu 
>You mean the different length of the bytearray(...?

No, I'm revering to the amount of lines. It switches between 1 - 4 lines in the same run instance. Those screenshots are took milliseconds from each other.

**Comment by WK-GiHu** [comment]
@JohannesAppel 
>Traceback (most recent call last):
  File "C:\Users\Hex\eclipse-workspace\DataLogger\src\canbus.py", line 122, in <module>
    cbus = CanBus(channel=0)
TypeError: __init__() missing 1 required positional argument: 'cnt'

Should read
```
self.cnt = self.counter()
```
`.counter(...` is a class method.

---

>No, I'm revering to the amount of lines. It switches between 1 - 4 lines in the same run instance. 

Is this not related to: ***There is an unknown number of data lines.***
Wait until, `.counter()` will work.

**Comment by JohannesAppel** [comment]
@WK-GiHu 
It worked, here is the output:
```
canlib version: 8.30
Using channel: Kvaser Leaf Professional HS (channel 0), EAN: 73-30130-00404-7
CanBus.stop()
Trashed frames:6008
EXIT, frame_count:7008
```

I ran the code again, the frame count and thrashed frames always differ with exactly a `1000`.

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>Trashed frames:6008
EXIT, frame_count:7008

Can't believe that result (6008), ok last try, disable Line 131
https://github.com/JohannesAppel/johannesappel-/blob/d4862f86f6972daef5eb7b69b2a1a37003e10652/DataLogger/src/canbus.py#L131

Do you know how many Frames your Hardware fires?

**Comment by JohannesAppel** [comment]
@WK-GiHu 
When the `time.sleep` is disabled the program does't run long. It doesn't show any errors but it stops running after 2-5 frames.

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>When the time.sleep is disabled the program does't run long. It doesn't show any errors but it stops running after 2-5 frames.

This indicates the `CanBus.run()` using `canlib.read()` is **uncooperative**.
Means, it grabs all **Timeslices** so the `__main__` Thread never get a **Timeslice** to run.

Force the `CanBus.run()` to be **cooperative** by insert before Line 53
https://github.com/JohannesAppel/johannesappel-/blob/0b6832fd40a0bc7757494809f35f34baa983fc09/DataLogger/src/canbus.py#L53
```
time.sleep(0.005)
```
Enable Line 131 again
https://github.com/JohannesAppel/johannesappel-/blob/d4862f86f6972daef5eb7b69b2a1a37003e10652/DataLogger/src/canbus.py#L131

---

Now both Thread using the same value `0.005`, you must find a balance that the **producer**, here `CanBus.run()=>queue.put()`, is minimal slower as the **consumer**, here `cbus.get()`.

**Comment by JohannesAppel** [comment]
@WK-GiHu 
It works a lot better now.
```
canlib version: 8.30
Using channel: Kvaser Leaf Professional HS (channel 0), EAN: 73-30130-00404-7
CanBus.stop()
Trashed frames:1
EXIT, frame_count:999
```
```
Using channel: Kvaser Leaf Professional HS (channel 0), EAN: 73-30130-00404-7
CanBus.stop()
Trashed frames:8
EXIT, none_count:10464
```

**Comment by WK-GiHu** [comment]
@JohannesAppel 

>It works a lot better now.

What about **real time**?

>EXIT, none_count:10464

Missing `frame_count:???` ?

**Comment by JohannesAppel** [comment]
@WK-GiHu 
>What about real time?

It's not really possible to say at the moment since the amount of data lines that's being displayed is still not consistent. It still "flashes", don't know how else to describe it. 
Referring to my `LogData` module, that's why I'm using a list in my code so that the whole list will be displayed each time I want to display something on `tkinter`, even if a specific data line wasn't read at that moment. It's because some data lines are read every 100ms and other 500ms. That was my work around the fact the data lines is inconsistent. 

>Missing frame_count:??? ?

Sorry, was running `main.py` not `canbus.py`.
```
canlib version: 8.30
Using channel: Kvaser Leaf Professional HS (channel 0), EAN: 73-30130-00404-7
CanBus.stop()
Trashed frames:1
EXIT, frame_count:969
```
```
canlib version: 8.30
Using channel: Kvaser Leaf Professional HS (channel 0), EAN: 73-30130-00404-7
CanBus.stop()
Trashed frames:1
EXIT, frame_count:909
```

**Comment by WK-GiHu** [comment]
@JohannesAppel 

><What about real time?
>It still "flashes",

I meant running `canbus.py __main__`.

**Comment by JohannesAppel** [comment]
@WK-GiHu 
>I meant running canbus.py __main__.

The only real way to tell if the data is real time is when I connect it to a bike, and for example turn the lights on the specific byte is supposed to change instantly. I don't know of another way to tell if it's real time.
